### PR TITLE
feature/tab-expand-pwsh-alias

### PR DESCRIPTION
**Pull request by csc027**
This patch implements what @jkymarsh wanted in #257.  I also included changes to handle the newer expansion logic with `Register-ArgumentCompleter`.  Many of the basic test cases for the proxy commands have already been implemented, but I think I may be missing some edge cases here.  In broader terms, what am I missing in terms of test cases?

There is also the issue of the regular expression not not handling all whitespace characters.  It currently handles the case of multiline commands.  For example:
```powershell
function Test {
    git `
    checkout `
    $args
}
```
The original regular expression used the `\s` character class, 
```powershell
"(^|[|;])\s*(?<cmd>$(Get-AliasPattern git).* \`$args)\s*($|[|;])"
```
but incorrectly allowed multiline commands without the backticks.

The regular expression for reference:
```powershell
# The regular expression here is roughly follows this pattern:
#
# <begin anchor><whitespace>*<git>(<whitespace><parameter>)*<whitespace>+<$args><whitespace>*<end anchor>
#
# The delimiters inside the parameter list and between some of the elements are non-newline whitespace characters ([^\S\r\n]).
# In those instances, newlines are only allowed if they preceded by a non-newline whitespace character.
#
# Begin anchor (^|[;`n])
# Whitespace   (\s*)
# Git Command  (?<cmd>$(GetAliasPattern git))
# Parameters   (?<params>(([^\S\r\n]|[^\S\r\n]``\r?\n)+\S+)*)
# $args Anchor (([^\S\r\n]|[^\S\r\n]``\r?\n)+\`$args)
# Whitespace   (\s|``\r?\n)*
# End Anchor   ($|[|;`n])
$script:GitProxyCommandRegex = "(^|[;`n])(\s*)(?<cmd>$(Get-AliasPattern git))(?<params>(([^\S\r\n]|[^\S\r\n]``\r?\n)+\S+)*)(([^\S\r\n]|[^\S\r\n]``\r?\n)+\`$args)(\s|``\r?\n)*($|[|;`n])"
```

### CODE REVIEW

**Comment by csc027** [comment]
Right now, there's an issue with how the proxy command expansion works in the PowerShell 6+ tab expansion implementation.  Because the registration for the proxy commands needs to happen at module load time, any changes to the `$Global:GitTabSettings` object after the module load will be ineffective.  There are two ways I can see to solve this:

1. Add a function that can register the proxy commands after module load.
2. Change the way the settings object is generated (i.e., do not overwrite an existing object and allow users to enable settings before the module is loaded).

I'm leaning towards the first option, but I realize it isn't very idiomatic for the module to configure settings via a function.

**Comment by MisinformedDNA** [comment]
@dahlbyk Where are we with this?

**Comment by csc027** [comment]
@MisinformedDNA I've been using this implementation for about six weeks on Windows with Powershell 7, and I haven't encountered any issues yet. It may fail on some cases I haven't encountered, but it shouldn't be too hard to push fixes in with the tests I added.

Still waiting on the review though.

**Comment by rkeithhill** [comment]
Since this isn't enabled by default, I think we should add this for v1 and have folks try it out.  I'll be adding some review comments here shortly mostly about style nits.  After those are addressed, I'll accept this PR.

File: src/GitTabExpansion.ps1 [hunk 1]
```diff
@@ -472,6 +488,34 @@ function GitTabExpansionInternal($lastBlock, $GitStatus = $null) {
...
 
+function Expand-GitProxyCommand($Command) {
+    # Make sure the incoming command matches: <Command> <Args>, so we can extract the alias/command
+    # name and the arguments being passed in.
+    if ($Command -notmatch '^(?<command>\S+)([^\S\r\n]|[^\S\r\n]`\r?\n)+(?<args>([^\S\r\n]|[^\S\r\n]`\r?\n|\S)*)$') {
+        return $Command
+    }
+
+    # Store arguments for replacement later
+    $Arguments = $Matches['args']
```
Comment (rkeithhill) [inline src/GitTabExpansion.ps1:499]: Local vars should be lower-case e.g `arguments`.  Also use `$matches['args']`.  We are mostly consistent on using $matches vs $Matches but I see there are three spots still using $Matches. Let's not add more.  :-)

Comment (csc027) [reply src/GitTabExpansion.ps1]: Makes sense.

File: src/GitTabExpansion.ps1 [hunk 2]
```diff
@@ -472,6 +488,34 @@ function GitTabExpansionInternal($lastBlock, $GitStatus = $null) {
     }
 }
 
+function Expand-GitProxyCommand($Command) {
```
[the remark this answers is not in the document]
Comment (csc027) [reply src/GitTabExpansion.ps1]: Sounds good to me.  I'll switch to `Expand-GitProxyFunction`.

File: src/GitTabExpansion.ps1 [hunk 3]
```diff
@@ -69,6 +69,22 @@ $script:gitCommandsWithParamValues = $gitParamValues.Keys -join '|'
...
+# In those instances, newlines are only allowed if they preceded by a non-newline whitespace character.
+#
+# Begin anchor (^|[;`n])
+# Whitespace   (\s*)
+# Git Command  (?<cmd>$(GetAliasPattern git))
+# Parameters   (?<params>(([^\S\r\n]|[^\S\r\n]``\r?\n)+\S+)*)
+# $args Anchor (([^\S\r\n]|[^\S\r\n]``\r?\n)+\`$args)
+# Whitespace   (\s|``\r?\n)*
+# End Anchor   ($|[|;`n])
+$script:GitProxyCommandRegex = "(^|[;`n])(\s*)(?<cmd>$(Get-AliasPattern git))(?<params>(([^\S\r\n]|[^\S\r\n]``\r?\n)+\S+)*)(([^\S\r\n]|[^\S\r\n]``\r?\n)+\`$args)(\s|``\r?\n)*($|[|;`n])"
```
[the remark this answers is not in the document]
Comment (csc027) [reply src/GitTabExpansion.ps1:86]: I've switched to `$script:GitProxyFunctionRegex`.

File: src/posh-git.psm1 [hunk 4]
```diff
@@ -1,4 +1,4 @@
-param([bool]$ForcePoshGitPrompt, [bool]$UseLegacyTabExpansion)
+param([bool]$ForcePoshGitPrompt, [bool]$UseLegacyTabExpansion, [bool]$EnableProxyCommandExpansion)
```
[the remark this answers is not in the document]
Comment (csc027) [reply src/posh-git.psm1:1]: I've switched to using `$EnableProxyFunctionExpansion`.

File: src/GitTabExpansion.ps1 (hunk 2 above)
[the remark this answers is not in the document]
Comment (csc027) [reply src/GitTabExpansion.ps1]: I've also renamed the test file and the test file's contents to reflect the general change from `*ProxyCommand*` to `*ProxyFunction*`.

**Comment by csc027** [comment]
Thanks @rkeithhill for reviewing.  😃

**Comment by MisinformedDNA** [comment]
@csc027 How can I enable this functionality?

**Comment by csc027** [comment]
@MisinformedDNA If you have the version with the change in, you have to import while setting the module arguments.  I don't think the PowerShell Gallery version has been updated with these changes yet; so, you'll likely need to import it from your Git repository directly.  Either way, the module arguments are:
```powershell
[bool]$ForcePoshGitPrompt,
[bool]$UseLegacyTabExpansion,
[bool]$EnableProxyFunctionExpansion
```
, which can be set via:
```powershell
Import-Module -Name '<path to posh-git.psd1 file>' -ArgumentList @($false, $false, $true)
```

File: src/GitTabExpansion.ps1 [hunk 5]
```diff
@@ -69,6 +69,22 @@ $script:gitCommandsWithParamValues = $gitParamValues.Keys -join '|'
...
+# In those instances, newlines are only allowed if they preceded by a non-newline whitespace character.
+#
+# Begin anchor (^|[;`n])
+# Whitespace   (\s*)
+# Git Command  (?<cmd>$(GetAliasPattern git))
+# Parameters   (?<params>(([^\S\r\n]|[^\S\r\n]``\r?\n)+\S+)*)
+# $args Anchor (([^\S\r\n]|[^\S\r\n]``\r?\n)+\`$args)
+# Whitespace   (\s|``\r?\n)*
+# End Anchor   ($|[|;`n])
+$script:GitProxyFunctionRegex = "(^|[;`n])(\s*)(?<cmd>$(Get-AliasPattern git))(?<params>(([^\S\r\n]|[^\S\r\n]``\r?\n)+\S+)*)(([^\S\r\n]|[^\S\r\n]``\r?\n)+\`$args)(\s|``\r?\n)*($|[|;`n])"
```
Comment (rkeithhill) [inline src/GitTabExpansion.ps1:86]: I wonder if we should define this regex in the GitTabSettings at the tops of this script:
```
$Global:GitTabSettings = New-Object PSObject -Property @{
    AllCommands = $false
    KnownAliases = @{
        '!f() { exec vsts code pr "$@"; }; f' = 'vsts.pr'
    }
    EnableLogging = $false
    LogPath = Join-Path ([System.IO.Path]::GetTempPath()) posh-git_tabexp.log
}
```
If that regex didn't work for a particular user's proxy function, they could then tweak the regex without requiring a turn of posh-git.

**Review by rkeithhill — COMMENTED** [review]
This LGTM.  Only issue is that I think perhaps the regex should be moved to the GitTabSettings object so that folks have a chance to tweak the regex for their proxy functions.

**Comment by rkeithhill** [comment]
Oops.  This is already merged.  Yikes this tab has been hanging around in my browser for a long time.  :-)

File: src/GitTabExpansion.ps1 (hunk 5 above)
Comment (csc027) [reply src/GitTabExpansion.ps1:86]: This sounds like a good idea, and it should work for the legacy and PowerTab implementations.  However, I don't think it would help for the `Register-ArgumentCompleter` implementation.  We would need to set the regular expression at module load time for the setting to have any effect.  Maybe there should be some sort of refactoring to allow settings to be set before the module is loaded.  Maybe allow the user to set a hash table that is marshalled into settings type at module load?

**Comment by SB-USNC** [comment]
This seems to work on aliases but not on functions. For example: 

```
function Get-GitSwitch { & git switch -v $args }
New-Alias -Name gw -Value Get-GitSwitch -Force -Option AllScope
```

`gw` does provide a Menu Complete  with Ctrl + Space 

still have to write `g switch`

**Comment by kleinfreund** [comment]
@SB-USNC The tab completion should work for the function. What doesn't work here is the tab completion for the alias to the function.

I use a lot of functions defined like this and with this feature enabled, I get tab completion (and Ctrl+Space) when using `gsw`.

```powershell
function gsw { git switch $args }
```
